### ISSUE TITLE
FlutterError.onError not catching exceptions produced in the app, as part of the SDK

### ISSUE THREAD

**Issue by jorgecoca** [flutter/flutter#48972]
## Steps to Reproduce

1. Setup your app inside `runZoned`
2. Setup `FlutterError.onError`
3. Throw an exception from a button tap

**Expected**: exception would be captured by `FlutterError.onError`
**Actual**: exception is printed on console, but not captured by `FlutterError.onError`

### Flutter Doctor

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel beta, v1.12.13+hotfix.6, on Mac OS X 10.15.2 19C57, locale en-US)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.6)
[✓] VS Code (version 1.41.1)
[✓] Connected device (1 available)

• No issues found!
```

### Sample

```main.dart
import 'dart:async';
import 'package:flutter/material.dart';

Future<void> main() async {
  FlutterError.onError = (FlutterErrorDetails errorDetails) {
    print('This is an error on the Flutter SDK');
    print(errorDetails.exception);
    print('-----');
    print(errorDetails.stack);
  };
  runZoned(() {
    runApp(App());
  }, onError: (error, stackTrace) {
    print('This is a pure Dart error');
    print(error);
    print('-----');
    print(stackTrace);
  });
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: RaisedButton(
            child: Text('Throw'),
            onPressed: () {
              throw Exception('my error');
            },
          ),
        ),
      ),
    );
  }
}
```

**Comment by julianbraasch** [comment]
tl;dr: Your code to catch exceptions is probably right, but I don't know why it didn't work when throwing the exception directly.

I have taken your code and rewrote it so that it doesn't throw an exception directly. Instead, I try to await a future but set a timeout on it which is too small.

I changed the line

```dart
onPressed: () {
  throw Exception('my error');
},
```

to

```dart
onPressed: () async {
  final result = await longFunction().timeout(Duration(seconds: 5));
  print(result);
}
```

```longFunction()``` is just a function that sleeps 10 second and then returns an arbitrary value:

```dart
Future<String> longFunction() async {
  await Future.delayed(Duration(seconds: 10), () {});
  return "Value";
}
```
Note how ```longFunction()``` would normaly take 10 seconds to complete whereas I only allow a timeout of 5 seconds.

Output (as expected):
```
I/flutter ( 6290): This is a pure Dart error
I/flutter ( 6290): TimeoutException after 0:00:05.000000: Future not completed
I/flutter ( 6290): -----
I/flutter ( 6290): #0      App.build.<anonymous closure> (package:flutter_trans/main.dart:35:40)
I/flutter ( 6290): <asynchronous suspension>
I/flutter ( 6290): #1      _InkResponseState._handleTap (package:flutter/src/material/ink_well.dart:654:14)
I/flutter ( 6290): #2      _InkResponseState.build.<anonymous closure> (package:flutter/src/material/ink_well.dart:729:32)
I/flutter ( 6290): #3      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:182:24)
I/flutter ( 6290): #4      TapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:365:11)
I/flutter ( 6290): #5      TapGestureRecognizer.handlePrimaryPointer (package:flutter/src/gestures/tap.dart:275:7)
I/flutter ( 6290): #6      PrimaryPointerGestureRecognizer.handleEvent (package:flutter/src/gestures/recognizer.dart:455:9)
I/flutter ( 6290): #7      PointerRouter._dispatch (package:flutter/src/gestures/pointer_router.dart:75:13)
I/flutter ( 6290): #8      PointerRouter.route (package:flutter/src/gestures/pointer_router.dart:102:11)
I/flutter ( 6290): #9      GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:218:19)
I/flutter ( 6290): #10     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:198:
```

**Comment by slightfoot** [comment]
There was a related issue floating about where if you make your main method async and call runZoned from it, it would not catch all uncaught exceptions.

To resolve it I had to await the runZoned which would return a future from runApp...
```dart
await runZoned(() async {
  runApp(App());
}, onError: (error, stackTrace) {
  //
});

This might be helpful.

**Comment by jorgecoca** [comment]
Unfortunately, that does not solve the issue (see GIF below). It should be hitting the breakpoints, or printing the custom log statements, but nothing...

![Screen-Recording-2020-01-17-at-12 27 07-PM](https://user-images.githubusercontent.com/3237451/72636601-fb8ab980-3924-11ea-84f5-69c156784a83.gif)

**Comment by jamie1192** [comment]
@jorgecoca I've run into the exact same issue, the only way i've gotten anything to trigger inside the `onError` callback is by awaiting a Future inside it-

```dart
runZoned<Future<Null>>(() async {
    LoggingHandler(environment: env);
    runApp(App());
  }, onError: (error, stackTrace) async {
    print("onError-main"); // this still never gets triggered, nor do breakpoints work
    await LoggingHandler.reportError(error, stackTrace);
  });
```

**Comment by fxsalazar** [comment]
Might be related #47447

**Comment by Pyroscout** [comment]
According to [this response](https://stackoverflow.com/a/57886463/4988151) to a similar question on stackoverflow, 

> It looks like when in debug mode, the flutter framework catches a lot of exceptions, prints to the console (and sometimes shows in the UI itself in red and yellow), but doesn't re-throw - so they are effectively swallowed and there's no way for your own code to catch them. But, when you deploy in release mode, this doesn't happen. So my minimal reproducible example does catch exceptions when built in release mode.

It seemed to work for me when I ran the app in `--release` mode.

**Comment by javierforero** [comment]
@jorgecoca  did you find and answer to this? Followed the entire thread and didn't find something that solved your issue. Experiencing the same thing.

**Comment by jorgecoca** [comment]
AFAIK it is still reproducible

**Comment by MisterJimson** [comment]
Just want to report that as of Flutter Beta 1.17.0 I am still reproducing this.

**Comment by erkkah** [comment]
It looks like this is happening since the `WidgetInspectorService` traps `onError` async, which in my case happens after I assign my own error handler.

A simple, stupid test (and a bit of framework debugging) verifies this:
```dart
  FlutterError.onError = _globalOnError;

  Timer.periodic(Duration(seconds: 1), (timer) {
    if (FlutterError.onError != _globalOnError) {
      print("Resetting global onError handler");
      FlutterError.onError = _globalOnError;
    }
  });
```

**Comment by erkkah** [comment]
Looks like this is a duplicate of https://github.com/flutter/flutter/issues/41540.

**Comment by alexobviously** [comment]
Just chiming in to say this is still a problem and prevents Sentry from working properly

**Comment by jamie1192** [comment]
@alexobviously , I commented a work around for Sentry/any error reporter in the [duplicate issue here](https://github.com/flutter/flutter/issues/41540#issuecomment-657958086).

It's an issue with Zone handling in debug mode.

**Comment by miquelbeltran** [comment]
I was facing the same issue, but when I tried in "release" mode on Android, I was able to see my custom logger capturing the Flutter errors, while when running on debug (from IntelliJ) my custom FlutterError.onError was being ignored. Maybe this info will help you!

**Comment by praveen-francium** [comment]
I am also facing the same issue the callback  onError is not being called at all.

**Comment by JamiLu** [comment]
Having run all the code that is in the `Crashy App` but in vain. Nothing really happens.

**Comment by erkkah** [comment]
Turning off "Flutter Structured Errors" in VSCode solves this for me and makes my custom error handler work.
Maybe there's a similar setting in other IDEs?

**Comment by Nico04** [comment]
> 
> 
> `FlutterError.onError` is not called if `WidgetsFlutterBinding.ensureInitialized();` function is called in the main. But the `ensureInitialized` might be needed in some cases where some resources are to be accessed before running the app itself. How this case should be handled?

I also call `WidgetsFlutterBinding.ensureInitialized();` in main, and it seems that I have the same issue : `FlutterError.onError` is not called.
Any solutions ?

**Comment by TahaTesser** [comment]
<details> 
<summary>code sample</summary> 

```dart
import 'dart:async';
import 'package:flutter/material.dart';

Future<void> main() async {
  FlutterError.onError = (FlutterErrorDetails errorDetails) {
    print('This is an error on the Flutter SDK');
    print(errorDetails.exception);
    print('-----');
    print(errorDetails.stack);
  };
  runZoned(() {
    runApp(App());
  }, onError: (error, stackTrace) {
    print('This is a pure Dart error');
    print(error);
    print('-----');
    print(stackTrace);
  });
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: RaisedButton(
            child: Text('Throw'),
            onPressed: () {
              throw Exception('my error');
            },
          ),
        ),
      ),
    );
  }
}

```

</details> 

<details> 
<summary>logs</summary> 

```bash
Launching lib\main.dart on Android SDK built for x86 in debug mode...
√ Built build\app\outputs\flutter-apk\app-debug.apk.
Connecting to VM Service at ws://127.0.0.1:50159/OdKo1a2e-fQ=/ws
D/EGL_emulation(18491): eglMakeCurrent: 0xe2f80060: ver 2 0 (tinfo 0xdaeadee0)
D/eglCodecCommon(18491): setVertexArrayObject: set vao to 0 (0) 1 0

[38;5;248m════════ Exception caught by gesture ═══════════════════════════════════════════[39;49m
[38;5;244mThe following _Exception was thrown while handling a gesture:[39;49m
Exception: my error

[38;5;244mWhen the exception was thrown, this was the stack[39;49m
[38;5;248m#0      App.build.<anonymous closure>[39;49m 
 package:master_flutter/main.dart
[38;5;244m#1      _InkResponseState._handleTap[39;49m 
 package:flutter/…/material/ink_well.dart
[38;5;244m#2      _InkResponseState.build.<anonymous closure>[39;49m 
 package:flutter/…/material/ink_well.dart
[38;5;244m#3      GestureRecognizer.invokeCallback[39;49m 
 package:flutter/…/gestures/recognizer.dart
[38;5;244m#4      TapGestureRecognizer.handleTapUp[39;49m 
 package:flutter/…/gestures/tap.dart
[38;5;244m...[39;49m
[38;5;244mHandler: "onTap"[39;49m
[38;5;244mRecognizer: TapGestureRecognizer#4c465[39;49m
    [38;5;244mdebugOwner: GestureDetector[39;49m
    [38;5;244mstate: possible[39;49m
    [38;5;244mwon arena[39;49m
    [38;5;244mfinalPosition: Offset(184.0, 370.9)[39;49m
    [38;5;244mfinalLocalPosition: Offset(31.6, 9.3)[39;49m
    [38;5;244mbutton: 1[39;49m
    [38;5;244msent tap down[39;49m
[38;5;248m════════════════════════════════════════════════════════════════════════════════[39;49m

```

</details> 


I can confirm works  https://github.com/flutter/flutter/issues/48972#issuecomment-689750210 
<details> 
<summary>logs</summary> 

```bash
Launching lib\main.dart on Android SDK built for x86 in debug mode...
√ Built build\app\outputs\flutter-apk\app-debug.apk.
Connecting to VM Service at ws://127.0.0.1:55515/LF7jtib9LWg=/ws
D/EGL_emulation(18291): eglMakeCurrent: 0xe2f805a0: ver 2 0 (tinfo 0xdae0f3c0)
D/eglCodecCommon(18291): setVertexArrayObject: set vao to 0 (0) 1 0
I/flutter (18291): This is an error on the Flutter SDK
I/flutter (18291): Exception: my error
I/flutter (18291): -----
[38;5;248mI/flutter (18291): #0      App.build.<anonymous closure>[39;49m 
 package:master_flutter/main.dart
[38;5;244mI/flutter (18291): #1      _InkResponseState._handleTap[39;49m 
 package:flutter/…/material/ink_well.dart
[38;5;244mI/flutter (18291): #2      _InkResponseState.build.<anonymous closure>[39;49m 
 package:flutter/…/material/ink_well.dart
[38;5;244mI/flutter (18291): #3      GestureRecognizer.invokeCallback[39;49m 
 package:flutter/…/gestures/recognizer.dart
[38;5;244mI/flutter (18291): #4      TapGestureRecognizer.handleTapUp[39;49m 
 package:flutter/…/gestures/tap.dart
[38;5;244mI/flutter (18291): #5      BaseTapGestureRecognizer._checkUp[39;49m 
 package:flutter/…/gestures/tap.dart
[38;5;244mI/flutter (18291): #6      BaseTapGestureRecognizer.handlePrimaryPointer[39;49m 
 package:flutter/…/gestures/tap.dart
[38;5;244mI/flutter (18291): #7      PrimaryPointerGestureRecognizer.handleEvent[39;49m 
 package:flutter/…/gestures/recognizer.dart
[38;5;244mI/flutter (18291): #8      PointerRouter._dispatch[39;49m 
 package:flutter/…/gestures/pointer_router.dart
[38;5;244mI/flutter (18291): #9      PointerRouter._dispatchEventToRoutes.<anonymous closure>[39;49m 
 package:flutter/…/gestures/pointer_router.dart
I/flutter (18291): #10     _LinkedHashMapMixin.forEach (dart:collection-patch/comp

```

</details> 


<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.22.0-10.0.pre.359, on Microsoft Windows [Version 10.0.19041.508], locale en-US)
    • Flutter version 1.22.0-10.0.pre.359 at C:\Code\flutter_master
    • Framework revision db6e2d8aa5 (2 hours ago), 2020-09-25 06:47:03 -0400
    • Engine revision 3a73d073c8
    • Dart version 2.11.0 (build 2.11.0-161.0.dev)


[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at C:\Code\sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = C:\Code\sdk
    • Java binary at: C:\Code\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.7.30503.244
    • Windows 10 SDK version 10.0.18362.0

[✓] Android Studio (version 4.0)
    • Android Studio at C:\Code\android-studio
    • Flutter plugin version 49.0.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] VS Code (version 1.49.2)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.14.1

[✓] Connected device (4 available)
    • Android SDK built for x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API 29) (emulator)
    • Windows (desktop)                  • windows       • windows-x64    • Microsoft Windows [Version 10.0.19041.508]
    • Web Server (web)                   • web-server    • web-javascript • Flutter Tools
    • Chrome (web)                       • chrome        • web-javascript • Google Chrome 85.0.4183.121

• No issues found!
```

</details>

**Comment by Nico04** [comment]
Same issue on Flutter stable v1.22
Any solutions yet ?

**Comment by erkkah** [comment]
Hi, @TahaTesser, @Nico04 !

This still only happens when running in a debugger, right?
I agree that the current behavior is unpredictable, since the attached debugger replaces the currently installed `onError` - handler some time after main is "complete".

But even if that was predictable - what should happen when both the debugger's (vscode in my case) and my own error handler is attached?

For me, turning off structured errors in vscode is fine, if I want to use my own error handler.
If I would like to use both, I would probably chain my handler to call the debugger handler as well, using a trick similar to https://github.com/flutter/flutter/issues/48972#issuecomment-645385209 to make sure my handler was the first responder.

So, as far as I see it, there are two viable workarounds to the problem, and nothing really blocking me.
Do you have other expectations here that I'm not thinking of? :)

**Comment by p02diada** [comment]
this is not working for me in debug or release mode, onError does not catch any Flutter SDK error

**Comment by Mike278** [comment]
From @julianbraasch's example:
> I changed the line
> 
> ```dart
> onPressed: () {
>   throw Exception('my error');
> },
> ```
> 
> to
> 
> ```dart
> onPressed: () async {
>   final result = await longFunction().timeout(Duration(seconds: 5));
>   print(result);
> }
> ```
> 

All I had to do was make the callback async and the error was no longer caught by `Flutter.onError`:
```dart
onPressed: () async {
  throw Exception('my error');
},
```
I guess this makes sense because `onPressed` isn't an async function so its an uncaught async exception, but I couldn't find any mention of `Flutter.onError` only catching synchronous exceptions.


To make this extra confusing, in our case the exception *also* wasn't landing at the zone's `onError` callback. This means even with both `onError` handlers configured to report uncaught exceptions, there are still uncaught and unreported exceptions.

```dart
Future<void> main() async {
  FlutterError.onError = (FlutterErrorDetails errorDetails) {
    print('This is an error on the Flutter SDK');
    print(errorDetails.exception);
    print('-----');
    print(errorDetails.stack);
  };

  WidgetsFlutterBinding.ensureInitialized(); // <- (1) called in a different zone as `runApp`
  runZoned(() {
    // WidgetsFlutterBinding.ensureInitialized(); // <- (2) called in a the same zone as `runApp`
    runApp(App());
  }, onError: (error, stackTrace) {
    print('This is a pure Dart error');
    print(error);
    print('-----');
    print(stackTrace);
  });
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: RaisedButton(
            child: Text('Throw'),
            onPressed: () async { // <- uncaught async exception
              throw Exception('my error');
            },
          ),
        ),
      ),
    );
  }
}
```

When only `(1)` is uncommented, the error does not land at either `onError` handler.
When only `(2)` is uncommented, the output is `This is a pure Dart error`

(related #50188).

**Comment by samandmoore** [comment]
I was experiencing this issue and I was able to achieve the expected behavior by doing this:

```dart
Future<void> main() async {
  runZonedGuarded(() {
     WidgetsFlutterBinding.ensureInitialized();

     FlutterError.onError = (FlutterErrorDetails errorDetails) {
       print('This is an error on the Flutter SDK');
       print(errorDetails.exception);
       print('-----');
       print(errorDetails.stack);
    };

    runApp(App());
  },
  (error, stackTrace) {
    print('This is a pure Dart error');
    print(error);
    print('-----');
    print(stackTrace);
  });
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          children: [
            RaisedButton(
              onPressed: () {
                Future.delayed(Duration.zero, () => throw Exception('async error'));
              },
              child: Text('async error'),
            ),
            RaisedButton(
              onPressed: () {
                throw Exception('sync error');
              },
              child: Text('sync error'),
            ),
          ],
        ),
      ),
    );
  }
}
```

it seems that moving the `WidgetsBinding.ensureInitialized()` and `FlutterError.onError` inside the `runZonedGuarded` call makes it so the following happens in debug builds (haven't been able to try a release build: 
* the synchronous error is reported via the `FlutterError.onError` path
* the asynchronous error is reported via the `runZonedGuarded` error callback

**Comment by drriguz** [comment]
got the same issue, any official reply?

**Comment by chvin** [comment]
My solution:
``` dart
void main() {
  FlutterError.onError = (FlutterErrorDetails details) async {
    // Send to Zone handle
    Zone.current.handleUncaughtError(details.exception, details.stack);
  };
  runZoned<Future<Null>>(() async {
    runApp(MyApp());
  }, onError: (error,  stackTrace) async {
    // You can handle error 
  });
}
```

**Comment by JustoSenka** [comment]
Having the same problem. Tried the solution above, didn't work, exceptions thrown from Futures are just swallowed, no log message no visual indicator on emulator. Not sure if it happens with all futures or only some, most of my code is async and I vaguely remember seeing exceptions as errors in my console before.

I have` WidgetsFlutterBinding.ensureInitialized(); ` in my main function cause I need it.

added this code below, doesn't seem to change anything:
```
runZonedGuarded<Future<void>>(() async {
      runApp(MyApp());
    }, (error, stackTrace) async {
      print(error.toString());
      throw error;
    });
```

**Comment by EArminjon** [comment]
In this example, the exception in not intercepted.
Real issue.

```dart
Future main() async {
  WidgetsFlutterBinding.ensureInitialized();

  runZonedGuarded(() {
    FlutterError.onError = (FlutterErrorDetails details) {
      print(details.exception);
      print(details.stack);
    };
    runApp(App());
  }, (Object error, StackTrace stack) {
    print(error);
    print(stack);
  });
}

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        brightness: Brightness.dark,
        backgroundColor: Colors.black,
      ),
      home: WillPopScope(
        onWillPop: () async {
          throw Exception("test");
          return false;
        },
        child: Scaffold(),
      ),
      debugShowCheckedModeBanner: false,
    );
  }
}
```

**Comment by nt4f04uNd** [comment]
@darshankawar could you close this as duplicate of https://github.com/flutter/flutter/issues/41540?

**Comment by TahaTesser** [comment]
@nt4f04uNd 
I have closed the other one in favor of this since this issue has more activity and higher upvotes
